src_c/rect.c

```diff
@@ -695,6 +695,142 @@ pg_rect_collidelistall(pgRectObject *self, PyObject *args)
     return ret;
 }
 
+static SDL_Rect *
+pgRect_FromObjectAndKeyFunc(PyObject *obj, PyObject *keyfunc)
+{
+    SDL_Rect temp;
+    if (keyfunc == NULL) {
+        return pgRect_FromObject(obj, &temp);
+    }
+
+    PyObject *obj_with_rect = PyObject_CallFunctionObjArgs(keyfunc, obj, NULL);
+    if (obj_with_rect == NULL) {
+        return NULL;
+    }
+
+    return pgRect_FromObject(obj_with_rect, &temp);
+}
+
+static PyObject *
+pg_rect_collideobjectsall(pgRectObject *self, PyObject *args, PyObject *kwargs)
+{
+    SDL_Rect *argrect;
+    Py_ssize_t size;
+    int loop;
+    PyObject *list, *obj;
+    PyObject *keyfunc = NULL;
+    PyObject *ret = NULL;
+    static char *keywords[] = {"list", "key", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "O|$O:collideobjectsall",
+                                     keywords, &list, &keyfunc)) {
+        return RAISE(PyExc_TypeError, "Could not parse args and kw.");
+    }
+
+    if (!PySequence_Check(list)) {
+        return RAISE(PyExc_TypeError,
+                     "Argument must be a sequence of objects.");
+    }
+
+    if (keyfunc == Py_None) {
+        keyfunc = NULL;
+    }
+
+    if (keyfunc != NULL && !PyCallable_Check(keyfunc)) {
+        return RAISE(PyExc_TypeError,
+                     "Key function must be callable with one argument.");
+    }
+
+    ret = PyList_New(0);
+    if (!ret) {
+        return NULL;
+    }
+
+    size = PySequence_Length(list); /*warning, size could be -1?*/
+    for (loop = 0; loop < size; ++loop) {
+        obj = PySequence_GetItem(list, loop);
+
+        if (!obj) {
+            Py_XDECREF(obj);
+            Py_DECREF(ret);
+            return RAISE(PyExc_TypeError, "Get item from sequence failed.");
+        }
+
+        if (!(argrect = pgRect_FromObjectAndKeyFunc(obj, keyfunc))) {
+            Py_XDECREF(obj);
+            Py_DECREF(ret);
+            return RAISE(
+                PyExc_TypeError,
+                "Item in sequence or returned by key function must be Rect.");
+        }
+
+        if (_pg_do_rects_intersect(&self->r, argrect)) {
+            if (0 != PyList_Append(ret, obj)) {
+                Py_DECREF(ret);
+                Py_DECREF(obj);
+                return NULL; /* Exception already set. */
+            }
+        }
+        Py_DECREF(obj);
+    }
+
+    return ret;
+}
+
+static PyObject *
+pg_rect_collideobjects(pgRectObject *self, PyObject *args, PyObject *kwargs)
```

Comment:
this functions has some mistakes, similar to the above function which I elaborated on in the reviews

Reply:
ok, corrected as hinted in the other method but now I get this when I run the tests:

```
======================================================================
ERROR: test_collideobjects_call_variants (__main__.RectTypeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Data\Programs\python390\lib\site-packages\pygame-2.1.3.dev3-py3.9-win-amd64.egg\pygame\tests\rect_test.py", line 2169, in test_collideobjects_call_variants
    self.assertRaises(TypeError, r.collideobjects, objects)
  File "C:\Data\Programs\python390\lib\unittest\case.py", line 731, in assertRaises
    return context.handle('assertRaises', args, kwargs)
  File "C:\Data\Programs\python390\lib\unittest\case.py", line 201, in handle
    callable_obj(*args, **kwargs)
SystemError: <built-in method collideobjects of pygame.Rect object at 0x00000220D0F0C480> returned NULL without setting an error

```

Reply:
this is resolved now